common/djangoapps/embargo/models.py

```diff
@@ -72,6 +82,213 @@ def embargoed_countries_list(self):
+        """
+        countries = cache.get(self.cache_key_name())
+        if not countries:
+            countries = list(RestrictedCourse.objects.values_list('course_key', flat=True))
+            cache.set(self.cache_key_name(), countries)
+        return countries
+
+    def __unicode__(self):
+        return unicode(self.course_key)
+
+    def save(self, *args, **kwargs):
+        """
+        Clear the cached value when saving a RestrictedCourse entry
+        """
+        super(RestrictedCourse, self).save(*args, **kwargs)
+        cache.delete(self.cache_key_name())
+
+    def delete(self, using=None):
+        super(RestrictedCourse, self).delete()
+        cache.delete(self.cache_key_name())
+
+
+class Country(models.Model):
+    """Representation of a country.
+
+    This is used to define country-based access rules.
+    There is a data migration that creates entries for
+    each country code.
+
+    """
+    country = CountryField(
+        db_index=True, unique=True,
+        help_text=ugettext_lazy(u"Two character ISO country code.")
+    )
+
+    def __unicode__(self):
+        return u"{name} ({code})".format(
+            name=unicode(self.country.name),
+            code=unicode(self.country)
+        )
+
+    class Meta:
+        # Default ordering is ascending by country code
+        ordering = ['country']
+
+
+class CountryAccessRule(models.Model):
+    """Course access rule based on the user's country.
+
+    The rule applies to a particular course-country pair.
+    Countries can either be whitelisted or blacklisted,
+    but not both.
+
+    To determine whether a user has access to a course
+    based on the user's country:
+
+    1) Retrieve the list of whitelisted countries for the course.
+    (If there aren't any, then include every possible country.)
+
+    2) From the initial list, remove all blacklisted countries
+    for the course.
+
+    """
+
+    RULE_TYPE_CHOICES = (
+        (WHITE_LIST, 'Whitelist (allow only these countries)'),
+        (BLACK_LIST, 'Blacklist (block these countries)'),
+    )
+
+    rule_type = models.CharField(
+        max_length=255,
+        choices=RULE_TYPE_CHOICES,
+        default=BLACK_LIST,
+        help_text=ugettext_lazy(
+            u"Whether to include or exclude the given course. "
+            u"If whitelist countries are specified, then ONLY users from whitelisted countries "
+            u"will be able to access the course.  If blacklist countries are specified, then "
+            u"users from blacklisted countries will NOT be able to access the course."
+        )
+    )
+
+    restricted_course = models.ForeignKey(
+        "RestrictedCourse",
+        help_text=ugettext_lazy(u"The course to which this rule applies.")
+    )
+
+    country = models.ForeignKey(
+        "Country",
+        help_text=ugettext_lazy(u"The country to which this rule applies.")
+    )
+
+    @classmethod
+    def is_course_embargoed_in_country_list(cls, course_id, country, rule_type):
```

Comment:
I'd suggest moving more of the application logic into class methods like these.  Why not something like:
```
def check_country_access(course_id, country_code):
    # ...
```

This method could:
1) Retrieve *both* the whitelist and the blacklist.
2) Calculate the list of all countries that have access.  (Consolidate the whitelist -- if provided -- and the blacklist).  This list can be cached, so future queries will require only one request to memcached.
3) Check if the `country_code` is in the list of countries that have access.